### PR TITLE
add .NET versions 4.6.2/4.7/4.7.1/4.7.2

### PR DESCRIPTION
**Pull request by dsyme**
This adds support for detecting .NET Versions  4.6.2/4.7/4.7.1/4.7.2, for example when building and for the default behaviour of `fsc code.fs` and F# Interactive.

### CODE REVIEW

File: src/fsharp/SimulatedMSBuildReferenceResolver.fs [hunk 1]
```diff
@@ -18,7 +18,7 @@ let internal SimulatedMSBuildResolver =
     { new Resolver with 
         member x.HighestInstalledNetFrameworkVersion() = 
             let root = x.DotNetFrameworkReferenceAssembliesRootDirectory
-            if Directory.Exists(Path.Combine(root,"v4.7.1")) then "v4.7.2"
+            if Directory.Exists(Path.Combine(root,"v4.7.2")) then "v4.7.2"
```
Comment (dsyme) [inline src/fsharp/SimulatedMSBuildReferenceResolver.fs]: Note there was a mistake here

**Comment by cartermp** [comment]
File | New project detects reference assembly folders and adds them as appropriate.

Not sure where that website is, but it's misleading. net462 and above are optional components, so you need to actually install them first before they become available.

**Comment by dsyme** [comment]
@KevinRansom @brettfo I don't know how to complete this, because ti seems like the MSBuild we are using is not up-to-date. The `TargetDotNetFrameworkVersion` enum is missing values for `Version47`, `Version471`, `Version472`.

Can we use an updated MSBuild?



```
C:\GitHub\dsyme\visualfsharp\src\fsharp\MSBuildReferenceResolver.fs(113,58): error FS0039: The field, constructor or me
mber 'Version47' is not defined. [C:\GitHub\dsyme\visualfsharp\src\fsharp\FSharp.Compiler.Private\FSharp.Compiler.Priva
te.fsproj]

C:\GitHub\dsyme\visualfsharp\src\fsharp\MSBuildReferenceResolver.fs(114,59): error FS0039: The field, constructor or me
mber 'Version471' is not defined. [C:\GitHub\dsyme\visualfsharp\src\fsharp\FSharp.Compiler.Private\FSharp.Compiler.Priv
ate.fsproj]

C:\GitHub\dsyme\visualfsharp\src\fsharp\MSBuildReferenceResolver.fs(115,59): error FS0039: The field, constructor or me
mber 'Version472' is not defined. [C:\GitHub\dsyme\visualfsharp\src\fsharp\FSharp.Compiler.Private\FSharp.Compiler.Priv
ate.fsproj]
```

**Comment by dsyme** [comment]
@cartermp Just to mention that there is some risk in this change  

Specifically, it would change F# scripts running with F# Interactive to use .NET 4.7.2 by default if it is installed.  Normally I’d say “go for it” but there is at least one case where we know this is a breaking change: the addition of some `Span<T>` overloads for `File.GetFullPath` and `File.GetFileName` means this code no longer resolves and needs a type annotation:

    let getFullPath a = Path.GetFullPath a
    let getFilename a = Path.GetFileName a

We’ve always done this – just bump the .NET version to make the latest the default – but in those cases we were normally shipping a new version of Visual Studio so users expected some transition. Perhaps we should now only do it at next VS given the above.  That will mean users of F# Interactive won't be able to access `Span<T>` or any .NET 4.6.2/4.7/4.7.1/4.7.2 built-ins until then though.

Separately, this again motivates having either `#r "framework: net472"`  to be able to pin down the default referenced framework.

**Comment by smoothdeveloper** [comment]
@dsyme, if the only issues we foresee would be the added overloads which need annotations, I'd recommend we bump the version regardless, it is better for users to fix their scripts to make them forward compatible while not excluding usage of latest apis available from FSI.

**Comment by cartermp** [comment]
@dsyme Yep, we can't change the default in dev15. We can reconsider this in dev16.

**Comment by cartermp** [comment]
@dsyme just as a note about span - since they won't be available in .NET Framework as per the current plan, the disambiguation may not be needed if scripts executing against .NET Framework. They will be needed for those executing against .NET Core, so that will have to factor into the design that @KevinRansom has put together and the RFC updates we'll make. At least I think it will...

**Comment by cartermp** [comment]
Also we don't know what the base version of .NET Framework for VS 2019 will be. We're currently working on negotiating a high enough version such that .NET Standard 2.0 components are "natively" understood, but if they are not, it is likely that FSI will be tied to that base version just so that the lowest-targetable projects (without further configuration) will work out of the box.

**Comment by cartermp** [comment]
I don't think this will be needed, but I could be wrong. For dev16, we'll have net471 as a base dependency, which means it will "natively" understand .NET Standard components. That's made a lot of us happy around here.

**Comment by KevinRansom** [comment]
@cartermp, 

I think this is may still be important, the scenario is when you are running FSC and building a project targeting one of these extra frameworks.

However, I made some changes when I fixed the msbuild dependency, and so this may not be necessary from that point of view.  I'm pretty sure I removed the use of the registry for example.  Anyway, I will dig into this a bit, when I get a moment.

**Comment by KevinRansom** [comment]
Handling this with this PR: #6005

**Pull request closed without merging**